File: NativeCounter/android/app/src/main/java/com/nativecounter/CounterManager.kt
package com.nativecounter

import com.facebook.react.common.MapBuilder
import com.facebook.react.uimanager.SimpleViewManager
import com.facebook.react.uimanager.ThemedReactContext
import com.facebook.react.uimanager.annotations.ReactProp

class CounterManager : SimpleViewManager<CounterView>() {
    // 리액트 네이티브 컴포넌트의 이름을 결정하는 메서드
    override fun getName(): String {
        return REACT_CLASS
    }

    override fun createViewInstance(reactContext: ThemedReactContext): CounterView {
        return CounterView(reactContext)
    }

    // 이벤트 연동
    override fun getExportedCustomBubblingEventTypeConstants(): MutableMap<String, Any> {
        val builder = MapBuilder.builder<String, Any>()
        return builder
            .put(
                // pressLeftButton 이벤트를 받으면 onPressLeftButton 호출
                "pressLeftButton", MapBuilder.of(
                    "phasedRegistrationNames", MapBuilder.of(
                        "bubbled", "onPressLeftButton"
                    )
                )

                // 위 내용은 사실상 다음 객체와 같다.
                // { pressLeftButton: { phasedRegistrationNames: bubbled: { onPressLeftButton } } }
            )
            .put(
                "pressRightButton", MapBuilder.of(
                    "phasedRegistrationNames",
                    MapBuilder.of("bubbled", "onPressRightButton")
                )
            )
            .build()
    }

    @ReactProp(name = "leftButtonText")
    fun setLeftButtonText(view: CounterView, text: String) {
        view.setLeftButtonText(text)
    }

    @ReactProp(name = "rightButtonText")
    fun setRightButtonText(view: CounterView, text: String) {
        view.setRightButtonText(text)
    }

    @ReactProp(name = "value")
    fun setValue(view: CounterView, value: Int) {
        view.setValue(value)
    }

    companion object {
        const val REACT_CLASS = "Counter"
    }
}

File: NativeCounter/ios/NativeCounter/CounterManager.m
#import <React/RCTViewManager.h>


// 네이티브 컴포넌트를 리액트에 제공
// 클래스 이름인 "<Counter />" (Manager)로 컴포넌트 명이 결정됨
// props 연동은 여기서 진행
@interface RCT_EXTERN_MODULE(CounterManager, RCTViewManager)

// props명, 타입
RCT_EXPORT_VIEW_PROPERTY(value, NSNumber)
RCT_EXPORT_VIEW_PROPERTY(leftButtonText, NSString)
RCT_EXPORT_VIEW_PROPERTY(rightButtonText, NSString)

RCT_EXPORT_VIEW_PROPERTY(onPressLeftButton, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onPressRightButton, RCTDirectEventBlock)

@end
